Page/fees explainer for users (#58)
ethereum-optimism/community-hub: src/docs/users/fees-in-optimistic-ethereum.md

:::

This document is for users of Optimistic Ethereum (OE), including developers and general end-users that seek to learn more about how _fees_ work in OE.
We hope that the content here can serve as a reference for understanding most of what you need to know about fees in OE.

_Note: if at any time in this process you get stuck or have questions, please reach out on discord! We just ask that you do this in the public #tech-support channel so that others may learn alongside us._ 😃

## Overview

We'll be primarily covering gas costs and savings experienced in _and_ around OE when you:

1. Transact on L2
2. Initiate an L1 Deposit
3. Complete a withdrawal (i.e., L2 → L1)

We'll also cover how gas costs are calculated and what parameters to optimize for to ensure that your contract deployments and transactions don't consume too much gas on L2.

::: tip L2 == OE, L1 == Ethereum
For the rest of this document, whenever we refer to L2, unless otherwise stated, we mean Optimistic Ethereum. Similarly, whenever we refer to L1, we mean the Ethereum mainnet.
:::

So, let's learn about gas in OE!

## 💵 1. Transacting on L2

### How are fees calculated on L2?

> Gas fees on L2 are charged in ETH (remember, [L2 ETH is not the same as ETH on L1](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth)) 

Reviewer: This should be https. 
Please use the relative linking format (ie. https://github.com/ethereum-optimism/community-hub/blob/159432ef32b0783422d4961ee9ba56c8df018999/src/docs/protocol/protocol.md#L49). That should also ensure https is used.

> However, **the gas fee payment itself is equal to to your `gasLimit * gasPrice`**, a calculation that is slightly different from L1 where the fee you are charged is based on `gasUsed`.

Reviewer: I know I wanted to be more concise, but if the audience is users here, we should define each of these values.

Due to this slight difference, **we recommend using `eth_estimateGas` to calculate your `gasLimit`.**
This is because `gasPrice` is a constant value of `1 Gwei`, so your focus should be on optimizing the `gasLimit`.

Two critical points to remember for this section here are:

> 1. you should never drop your `eth_gasPrice` or your transaction will simply get rejected on L2, and

Reviewer: I think `reduce` is a better term than `drop`. Also 'drop from' what?

2. never drop your `gasLimit` below the value returned by `eth_estimateGas` `gasLimit`, or your transaction might revert.

### 👩‍💻 Transactions for developers

As a developer, it's important to know that `eth_estimateGas` returns:

```

> rollupTxSize * dataPrice + executionPrice * gasUsed

Reviewer: Same for these valued: `rollupTxSize`, `dataPrice`, `executionPrice`

```

(See [ethereum-optimism/go-ethereum/pull/273](https://github.com/ethereum-optimism/go-ethereum/pull/273) for more info on these variables.)

**To minimize gas costs for users, we recommend minimizing your `rollupTxSize`.**
For example, you can reduce your `uint256` state variables to `uint32` variables wherever possible in your transaction inputs.

> By reducing the size of your state variables, you will likely cut costs to reduce your `rollupTxSize` in exchange for increasing the amount of `gasUsed` for your transaction.

Reviewer: State variables or input variables? I'd assume the latter, but I'm not sure.

**Additionally, you can expect the `executionPrice` to be fairly minimal, unless L2 becomes congested.**

### What is the gas limit on L2?

Some final notes on the amount of `gasUsed` per transaction on L2.

* `gasUsed` must always be under 9 million per transaction.
* `gasPrice` is fixed as a constant (it is initially set to `1 Gwei`).
* `gasUsed` is always equal to the `gasLimit` (since 100% of gas is always used).
* `executionGasUsed` is added as a new field in the transaction receipt.

**There is a per transaction gas limit of 9 million on L2 and when you compile your contracts for L2, gas usage can increase by more than 5x**.

So, if your transaction consumes more than 1.5 million gas on L1, you should really be mindful of the amount of gas it consumes on L2.

## 🏦 2. Initiating an L1 Deposit

The most immediate thing you'll likely want to do is deposit ether (ETH) to L2 so that you can interact with contracts and applications while on L2.

> (Note, [ETH on L2 is actually WETH](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth), but we abstract this for you 🙂.)

Reviewer: please use relative link format.

> (Note, [ETH on L2 is actually WETH](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth), but we abstract this for you 🙂.)

Reviewer: ```suggestion
(Note, [ETH on L2 is actually WETH](http://community.optimism.io/docs/protocol/evm-comparison.html#native-weth), but we abstract this away for you 🙂.)
```

Currently, the only way to do this is to initiate a deposit from L1 to L2.

### What can you expect to pay in gas costs when depositing from L1 to L2?

While your _real_ gas costs on OE will vary (depending on several factors like network demand or network congestion of the OE network, etc.), we can provide a _best estimate_ for what your expected costs will be.

First, let's see what the gas cost is for depositing SNX to L2.
[Open up this transaction on Etherscan](https://etherscan.io/tx/0xbc86558426c2c62fd49a57b830182b37c9e71646e1ba12aebf71b356253e785c).

If we look at the `Transaction Action`, we see a deposit was made for roughly `212.35 SNX` to the Synthetix L2 Bridge. Although this field gives us an easy look at the gas cost, _**it's very important to remember that gas costs (i.e. transaction fees) are dependent on the changing gas price and gas used by the transaction!**_
What this means is that you should really be mindful of the `Gas Price` in Gwei and the `Gas Used by Transaction` for your transactions.

> Next, if we look at the `Transaction Fee`, we can see that **this sender paid roughly `$78.42` for an SNX deposit to L2, a gas cost that you would expect for an L1 deposit.**

Reviewer: ```suggestion
Next, if we look at the `Transaction Fee`, we can see that **this sender paid roughly `$78.42` for an SNX deposit to L2, a gas cost that you would expect for a somewhat complex L1 transaction.**
```

# How Fees Work in Optimistic Ethereum

## Introduction

::: tip Work in Progress™
_Our documentation is a rapidly improving work in progress. If you have questions or feel like something is missing feel free to ask in our [Discord server](https://discord.gg/5TaAXGn2D8) where we (and our awesome community) are actively responding, or [open an issue](https://github.com/ethereum-optimism/community-hub/issues) in the GitHub repo for this site._
:::

This document is mainly for users of Optimistic Ethereum (OE), including developers and general end-users that seek to learn more about how _fees_ work in OE.
We hope that the content here can serve as a reference for understanding most of what you need to know about fees in OE.

_Note: if at any time in this process you get stuck or have questions, please reach out on discord! We just ask that you do this in the public #tech-support channel so that others may learn alongside us._ 😃

## Overview

We'll be primarily covering gas costs and savings experienced in _and_ around OE when you:

1. Initiate an L1 Deposit

> 2. Complete a withdrawal (i.e. L2 → L1)

Reviewer 2: ```suggestion
2. Complete a withdrawal (i.e., L2 → L1)
```
https://english.stackexchange.com/questions/6894/should-there-be-a-comma-after-i-e

This can add additional complexity to newcomers to dApps that only know and understand ETH but not WETH.

This is what OE solves by making ETH an ERC-20 on L2.
More info can be found in our [FAQs](http://community.optimism.io/faqs/#does-optimistic-ethereum-use-eth-natively) and in our [contracts](https://github.com/ethereum-optimism/contracts/blob/master/contracts/optimistic-ethereum/OVM/predeploys/OVM_ETH.sol)
:::,,

So, **what can you expect to pay in gas costs when making an L1 to L2 deposit?**

While your _real_ gas costs on OE will vary (depending on several factors like network demand or network congestion of the OE network, etc.), we can provide a _best estimate_ for what your expected costs will be.

First, let's see what the gas cost is for depositing Synthetix's token, SNX, into L2 OE.
[Open up this transaction on Etherscan](https://etherscan.io/tx/0xbc86558426c2c62fd49a57b830182b37c9e71646e1ba12aebf71b356253e785c).

We'll highlight several important fields from the `Overview` section here that we'll expand on:

* **Interacted With (To):**  Tells us the contract that was interacted with in this transaction.

> * **Transaction Action**: A concise explanation of the overall action that is being made in this transasction. You can use this field as a quick summary of the transaction details, excluding the fee that was paid and timestamps.

Author: This is perfect, thanks!

Although this field gives us an easy look at the gas cost, _**it's very important to remember that gas costs (i.e. transaction fees) are dependent on the changing gas price and gas used by the transaction!**_
What this means is that you should really be mindful of the `Gas Price` in Gwei and the `Gas Used by Transaction` for your transactions.
These variables and the `Ether Price` also let you calculate expected gas costs yourself, letting you make estimates for future transactions that you intend to make.
For example, this transaction used `336,454` gas, at a gas price of `116 Gwei` (or `0.000000116 ether`) and an ether price of `$2,009.19`.
This gives us a gas cost of `336,454 * 0.000000116 * 2009.19 = $78.42` for this SNX deposit transaction.

<!--

To make this calculation easy for you, we've included calculator to compute gas cost!

INSERT CALCULATOR HERE

-->

Thus, **we can see that this sender paid roughly `$78.42` for an SNX deposit to L2 OE**, an amount that is equal to what you would expect for an L1 Ethereum gas cost for a deposit.

> ### Main cost of a deposit

Reviewer: This section seems like it's trying to be a "how to read etherscan" tutorial. I assume it's motivated by previous attempts to answer certain questions, but it feels like it goes beyond the scope. I'd try to make it more concise by asking: 
1. What question are we trying to answer?
2. What information do we need to present here? 
3. What knowledge should we take as prerequisite (and provide links for users to gain that knowledge if they don't have it).

Thus, there is gas price risk in choosing to post-pone your self-withdrawal past the 7-day window.

## Transacting on L2

In this section we'll briefly review how much transactions cost and how they are calculated.
If you think something is missing from this section (or anywhere else in this document!), make sure to reach out to us on [Discord](https://discord.gg/5TaAXGn2D8) to let us know how we can improve this page!

### How are fees calculated in L2 Optimistic Ethereum?

To understand how much transaction fees cost on L2 OE, it behooves you to know how these fees are calculated.

As a gentle reminder, on L2 OE, gas fees are charged in ETH (but not the same ETH as in L1)!

However, the fee payment itself is equal to to your `gasLimit * gasPrice`, a calculation that is slightly different from L1 Ethereum where the fee you are charged is based on `gasUsed`.
Due to this slight difference, we recommend using `eth_estimateGas` to calculate your `gasLimit`.
This is because `gasPrice` is a constant value of `1 Gwei`, so your focus should be on optimizing the `gasLimit`.

> Two critical points to remember for this section here are: 1) you should never drop your `eth_gasPrice` or your transaction will simply get rejected on L2, and 2) never drop your `eth_estimateGas` `gasLimit` or your transaction might revert.

Reviewer: > never drop your `eth_estimateGas` `gasLimit` or your transaction might revert.

Do you mean, 'never drop `gasLimit` below the value returned by `eth_estimateGas`?

* `dataPrice` is a value set by the sequencer based on the current congestion on L1 (also known as "the cost of L1 data": if L1 gas prices are high, then `dataPrice is proportionally high).
* `executionPrice` is fetched via the standard `eth_gasPrice` rules that `geth` uses based on the current congestion on L2.

(See [ethereum-optimism/go-ethereum/pull/273](https://github.com/ethereum-optimism/go-ethereum/pull/273) for more info.)

From these gas estimation variables, this means that the majority of transaction costs until L2 becomes congested will be equal to `rollupTxSize * dataPrice`.
Additionally, you can expect for the `executionPrice` to be fairly minimal, unless L2 becomes congested.

To minimize gas costs for users, we recommend minimizing your `rollupTxSize`.
For example, you can reduce your `uint256` state variables to `uint32` variables wherever possible in your transaction inputs.
By reducing the size of your state variables, you will likely cut costs to reduce your `rollupTxSize` in exchange for increasing the amount of `gasUsed` for your transaction.

#### Final note on `gasUsed`

Some final notes on the amount of `gasUsed` per transaction on L2.

> * **The `gasUsed` must always be under 9 million**.

Reviewer: ```suggestion
* **`gasUsed` must always be under 9 million per transaction**.
```

Reviewer: Similarly remove "The" from the following lines

> * **The `gasPrice` is fixed as a constant (it is initially set to `1 Gwei`)**
> * **The `gasUsed` is always equal to the `gasLimit` (since 100% of gas is always used)**
> * Lastly, **we add `executionGasUsed` as a new field in the transaction receipt.

Reviewer: ```suggestion
* `gasPrice` is fixed as a constant (it is initially set to `1 Gwei`).
* `gasUsed` is always equal to the `gasLimit` (since 100% of gas is always used).
*  `executionGasUsed` is added as a new field in the transaction receipt.
```

No need to bold every line in a list.

> There is a per transaction gas limit of 9 million in L2 and when you compile your contracts for L2, gas usage can increase by more than 5x.

Reviewer: ```suggestion
There is a per transaction gas limit of 9 million on L2 and when you compile your contracts for L2, gas usage can increase by more than 5x.
```
We should be consistent in this phrasing

So, if your transaction does consume more than 1.5 million gas on L1, you should really be mindful of the amount of gas it consumes in L2.

## Advanced Technical Info

Here we present some more technical documentation on advanced subjects for those interested in learning more about gas from the level of the protocol.

### Nuisance Gas

> Nuisance gas is a separate subject from the kind the topics we covered above.

Reviewer: ```suggestion
Nuisance gas is a separate subject from the topics we covered above.
```

Nonetheless, this type of gas is included in L2 transaction costs.

So, what is _nuisance gas_?

> **Nuisance gas exists to **provide an upper bound on the _maximum possible L1 gas_ that a [_fraud proof_](https://research.paradigm.xyz/optimism#fraud-proofs) can cost.**

Reviewer: ```suggestion
Nuisance gas **provides an upper bound on the _maximum possible L1 gas_ that a [_fraud proof_](https://research.paradigm.xyz/optimism#fraud-proofs) can cost.**
```
